File: C++/Linked List/04_removing_loop_in_LL.cpp
/*
Idea:
Use  Floyd’s Cycle detection algorithm to detect if a loop is present in linked list
Once detected, finding length of the loop, and setting 2 pointers at head & other to k nodes after head
Incrementing both with same pace till both equal, ( we found the start of loop) & now finding last node & edit it
*/

//{ Driver Code Starts
// driver code

#include <bits/stdc++.h>
using namespace std;

struct Node
{
    int data;
    Node* next;

    Node(int val)
    {
        data = val;
        next = NULL;
    }
};

void loopHere(Node* head, Node* tail, int position)
{
    if (position == 0) return;

    Node* walk = head;
    for (int i = 1; i < position; i++)
        walk = walk->next;
    tail->next = walk;
}

bool isLoop(Node* head)
{
    if (!head) return false;

    Node* fast = head->next;
    Node* slow = head;

    while (fast != slow)
    {
        if (!fast || !fast->next) return false;
        fast = fast->next->next;
        slow = slow->next;
    }

    return true;
}

int length(Node* head)
{
    int ret = 0;
    while (head)
    {
        ret++;
        head = head->next;
    }
    return ret;
}

bool notOriginal(Node* head, unordered_map<Node*, int>& myMap) {

    while (head) {
        if (myMap.find(head) == myMap.end()) return true;
        if (myMap[head] != (head->data)) return true;

        head = head->next;
    }
}




class Solution
{
public:
    //Function to remove a loop in the linked list.
    void ifLoop(Node* node_in_loop, Node* head)
    {
        Node* ptr1 = node_in_loop->next, ptr2 = node_in_loop;
        int length = 1;

        //finding count of elements in loop 
        while (ptr1 != ptr2)
        {
            length++;
            ptr1 = ptr1->next;
        }

        ptr1 = head;
        for (int i = 0;i < length;++i)
            ptr2 = ptr2->next; // setting up 2 pointers (one at head & other to k nodes after head)

        while (ptr1 != ptr2)
        { //Move both pointers at the same pace, they will meet at loop starting node
            ptr1 = ptr1->next;
            ptr2 = ptr2->next;
        }

        while (ptr1 != ptr2->next)
        { // getting pointer to last node
            ptr2 = ptr2->next;
        }
        ptr2->next = NULL;
    }

    //function to check if there's loop in linked list
    void removeLoop(Node* head)
    {
        Node* front = head->next, * back = head;
        while (front != back)
        {
            if (front == NULL || front->next == NULL)
                return;
            front = front->next->next;
            back = back->next;
        }
        ifLoop(back, head);
    }
};

//{ Driver Code Starts.

int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        unordered_map<Node*, int>myMap;

        int n, num;
        cin >> n;

        Node* head, * tail;
        cin >> num;
        head = tail = new Node(num);

        myMap[head] = num;

        for (int i = 0; i < n - 1; i++)
        {
            cin >> num;
            tail->next = new Node(num);
            tail = tail->next;
            myMap[tail] = num;
        }

        int pos;
        cin >> pos;
        loopHere(head, tail, pos);

        Solution ob;
        ob.removeLoop(head);

        if (isLoop(head) || length(head) != n || notOriginal(head, myMap))
            cout << "0\n";
        else
            cout << "1\n";
    }
    return 0;
}

// } Driver Code Ends

File: Python/Linked List/04_removing_loop_in_LL.py
"""
Idea:
Use  Floyd’s Cycle detection algorithm to detect if a loop is present in linked list
Once detected, finding length of the loop, and setting 2 pointers at head & other to k nodes after head
Incrementing both with same pace till both equal, ( we found the start of loop) & now finding last node & edit it
"""


class Solution:
    # Function to remove a loop in the linked list.
    class Solution:
        # Function to remove a loop in the linked list.
        def ifLoop(self, node_in_loop, head):
            ptr1, ptr2 = node_in_loop.next, node_in_loop
            length = 1

            # Finding count of elements in the loop
            while ptr1 != ptr2:
                length += 1
                ptr1 = ptr1.next

            ptr1 = ptr2 = head
            for i in range(length):
                ptr2 = ptr2.next  # Setting up 2 pointers (one at head & other to k nodes after head)

            while ptr1 != ptr2:
                # Move both pointers at the same pace; they will meet at the loop starting node
                ptr1 = ptr1.next
                ptr2 = ptr2.next

            while ptr1 != ptr2.next:
                # Getting a pointer to the last node
                ptr2 = ptr2.next

            ptr2.next = None

        # Function to check if there's a loop in the linked list
        def removeLoop(self, head):
            front, back = head.next, head
            while front != back:
                if front is None or front.next is None:
                    return
                front = front.next.next
                back = back.next
            self.ifLoop(back, head)


# code here
# remove the loop without losing any nodes


# {
# Driver Code Starts
# driver code:

class Node:
    def __init__(self, val):
        self.next = None
        self.data = val


class linkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def add(self, num):
        if self.head is None:
            self.head = Node(num)
            self.tail = self.head
        else:
            self.tail.next = Node(num)
            self.tail = self.tail.next

    def isLoop(self):
        if self.head is None:
            return False

        fast = self.head.next
        slow = self.head

        while slow != fast:
            if fast is None or fast.next is None:
                return False
            fast = fast.next.next
            slow = slow.next

        return True

    def loopHere(self, position):
        if position == 0:
            return

        walk = self.head
        for _ in range(1, position):
            walk = walk.next
        self.tail.next = walk

    def length(self):
        walk = self.head
        ret = 0
        while walk:
            ret += 1
            walk = walk.next
        return ret


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        n = int(input())
        arr = tuple(int(x) for x in input().split())
        pos = int(input())

        ll = linkedList()
        for i in arr:
            ll.add(i)
        ll.loopHere(pos)

        Solution().removeLoop(ll.head)

        if ll.isLoop() or ll.length() != n:
            print(0)
            continue
        else:
            print(1)

# } Driver Code Ends
